assets/js/workflow-reducer.js

```diff
@@ -25,312 +41,567 @@ export function mockAPI(mock_api) {
-        selectedWfModule = action.workflow.wf_modules[0].id;
-      };
-      return Object.assign({}, state, {
-        selected_wf_module: selectedWfModule,
-        workflow: action.workflow
-      });
-
-
-    // Sets the selected module from backend if it exists, or the first module if there are any at all
-    case INITIAL_LOAD_WORKFLOW:
-      console.log("INITIAL_LOAD_WORKFLOW");
-      var selectedWfModule = null;
-      if (action.workflow.selected_wf_module) {
-        selectedWfModule = action.workflow.selected_wf_module;
-      } else if (action.workflow.wf_modules && action.workflow.wf_modules.length) {
-        selectedWfModule = action.workflow.wf_modules[0].id;
-      };
-      return Object.assign({}, state, {
-        selected_wf_module: selectedWfModule,
-        workflow: action.workflow
-      });
-
-
-    // Change id of module currently selected
-    case SELECTED_MODULE_CHANGE:
-      // console.log(SELECTED_MODULE_CHANGE + " old " +  state.selected_wf_module + " new " + action.id);
-      if (!'selected_wf_module' in state || (action.id != state.selected_wf_module)) {
-        // call API to set selected WF module in backend
-        api.setSelectedWfModule(state.workflow.id, action.id);
-        return Object.assign({}, state, {
-          selected_wf_module: action.id,
-        });
-      } else {
-        return state;
+  if (credentialIndex >= 0) {
+    return update(state, {
+      loggedInUser: {
+        google_credentials: { $splice: [[credentialIndex, 1]] }
       }
+    });
+  }
 
-    // Change status on a single module
-    case MODULE_STATUS_CHANGE:
-      //console.log(WF_MODULE_STATUS_CHANGE + " " + action.id);
-      if ('wf_modules' in state.workflow) {
-
-        var newState = state;
-
-        for (var wfm of newState.workflow.wf_modules) {
+  return state;
+});
 
-          // Find matching module, and change state if status changed or error message changed
-          if (wfm.id == action.id &&
-              ((wfm.status != action.status) || (wfm.status=='error' && wfm.error_msg != action.error_msg))) {
-            // console.log("actually changed status for " + wfm.id);
+// -- Workflow Module --
 
-            // Create a copy of the wf_module with new status
-            var newWfm = Object.assign({}, wfm, { status: action.status, error_msg: action.error_msg });
+// UPDATE_WF_MODULE
+// Patch a workflow module with new data
 
-            // Clone the state, switch out this one wfm (keep position in wf_modules array )
-            newState = Object.assign({}, state);
-            newState.workflow = Object.assign({}, state.workflow);
-            newState.workflow.wf_modules = state.workflow.wf_modules.map(el => el!==wfm ? el : newWfm);
-          }
+// TODO: We don't validate which fields or types are on
+// a WfModule here. The backend will reject nonexistent
+// fields, but should we do something on the frontend?
```

Comment:
I'm comfortable not checking, for the following reason: the backend MUST check, as a security concern. If the front end calls this with bogus fields it's a bug, but no harm done. However it does bring up the issue of run time assertions. We currently don't have any mechanism for them, and it's not clear what we would do with assertion failures if we did.